lib/chainweb/peer.ex

```diff
@@ -0,0 +1,54 @@
+defmodule Kadena.Chainweb.Peer do
+  @moduledoc """
+  `Peer` struct definition.
+  """
+  @behaviour Kadena.Chainweb.Type
+
+  @type address :: map() | nil
+  @type id :: String.t() | nil
+  @type error :: {:error, Keyword.t()}
+  @type result :: t() | error()
+  @type validation :: {:ok, map()} | error()
+
+  @type t :: %__MODULE__{
+          address: address(),
+          id: id()
+        }
+
+  defstruct [:id, :address]
+
+  @impl true
+  def new(attrs \\ nil)
+
+  def new(attrs) when is_list(attrs) do
+    id = Keyword.get(attrs, :id)
+    address = Keyword.get(attrs, :address)
+
+    %__MODULE__{}
+    |> set_id(id)
+    |> set_address(address)
+  end
+
+  def new(_attrs), do: %__MODULE__{}
+
+  @spec set_address(peer :: t(), address :: address()) :: result()
+  def set_address(%__MODULE__{} = peer, address) when is_map(address) do
```

Comment:
We could avoid the guard...
```suggestion
  def set_address(%__MODULE__{} = peer, %{} = address) do
```